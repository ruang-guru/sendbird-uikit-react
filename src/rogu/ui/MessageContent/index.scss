@import '../../../styles/variables';

$borderRadius: 0.5rem;
$bubbleCaretSize: 0.5rem;
$avatarSpacing: 0.5rem;
$avatarSize: 2rem;

.rogu-message-content {
  position: relative;
  display: flex;
  flex-direction: row;
  width: 100%;
  height: 100%;
  margin-bottom: 1rem;

  @include themed() {
    color: t(on-bg-1);
  }

  .rogu-message-content__avatar {
    margin-right: $avatarSpacing;
    width: $avatarSize;
    height: $avatarSize;
    flex-shrink: 0;
  }

  .rogu-message-content__content {
    display: flex;
    flex-direction: column;
    align-items: flex-end;
    position: relative;
    max-width: 35rem;

    .rogu-message-content__bubble {
      display: flex;
      flex-direction: column;
      padding: 0.75rem;
      border-top-left-radius: $borderRadius;
      border-top-right-radius: $borderRadius;
      border-bottom-left-radius: $borderRadius;
      border-bottom-right-radius: $borderRadius;

      .rogu-message-content__bubble__header {
        line-height: 0;
        display: flex;
        flex-wrap: nowrap;
        margin-bottom: 0.25rem;

        .rogu-message-content__sender-name {
          overflow: hidden;
          text-overflow: ellipsis;
          display: -webkit-box;
          word-break: break-all;
          -webkit-line-clamp: 1;
          -webkit-box-orient: vertical;
        }

        .rogu-message-content__operator-label {
          margin-left: 0.5rem;
          padding: 0.25rem 0.75rem;
          border-radius: 0.25rem;
          white-space: nowrap;

          @include themed() {
            color: t(on-bg-1);
            background-color: t(bg-1);
          }
        }

        .rogu-message-content__menu {
          flex-shrink: 0;
        }
      }

      &::before {
        content: '';
        width: 0;
        height: 0;
        border-style: solid;
        position: absolute;
        top: 0;
      }
    }

    .rogu-message-content__misc {
      margin-top: 0.25rem;
    }
  }

  &.rogu-message-content--incoming {
    justify-content: flex-start;

    .rogu-message-content__content {
      padding-left: $bubbleCaretSize;

      .rogu-message-content__bubble {
        border-top-left-radius: 0;

        @include themed() {
          background-color: t(bg-0);
        }

        &::before {
          border-width: 0 $bubbleCaretSize $bubbleCaretSize 0;
          left: 0;

          @include themed() {
            border-color: transparent t(bg-0) transparent transparent;
          }
        }
      }
    }
  }

  &.rogu-message-content--outgoing {
    justify-content: flex-end;

    .rogu-message-content__content {
      padding-right: $bubbleCaretSize;
      padding-left: $avatarSize + $avatarSpacing;

      .rogu-message-content__bubble {
        border-top-right-radius: 0;

        @include themed() {
          background-color: t(primary-1);
        }

        &::before {
          border-width: $bubbleCaretSize $bubbleCaretSize 0 0;
          right: 0;

          @include themed() {
            border-color: t(primary-1) transparent transparent transparent;
          }
        }
      }
    }
  }

  &.rogu-message-content--chain-bottom {
    margin-bottom: 0.5rem;
  }

  &.rogu-message-content--chain-top {
    .rogu-message-content__content {
      .rogu-message-content__bubble {
        &::before {
          display: none;
        }
      }
    }

    &.rogu-message-content--incoming {
      .rogu-message-content__content {
        padding-left: $avatarSize + $avatarSpacing + $bubbleCaretSize;

        .rogu-message-content__bubble {
          border-top-left-radius: 0;
        }
      }
    }

    &.rogu-message-content--outgoing {
      .rogu-message-content__content {
        padding-right: $bubbleCaretSize;

        .rogu-message-content__bubble {
          border-top-right-radius: 0;
        }
      }
    }
  }
}
