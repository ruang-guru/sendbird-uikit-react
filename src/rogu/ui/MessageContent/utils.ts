const colorSet = {
  "#DF4141": ["A", "B", "C", "D"],
  "#61CE5E": ["E", "F", "G", "H"],
  "#6073E2": ["I", "J", "K", "L"],
  "#F89825": ["M", "N", "O", "P"],
  "#2EB5C0": ["Q", "R", "S", "T"],
  "#BB58D0": ["U", "V", "W", "X"],
  "#00A5FF": ["Y", "Z"],
};

export const generateColorFromString = (str: string): string => {
  const firstChar = str[0] || "";
  const normalizedFirstChar = firstChar.toUpperCase();

  let color = "inherit";
  for (const [hex, chars] of Object.entries(colorSet)) {
    if (chars.includes(normalizedFirstChar)) {
      color = hex;
      break;
    }
  }

  return color;
};

export default generateColorFromString;
