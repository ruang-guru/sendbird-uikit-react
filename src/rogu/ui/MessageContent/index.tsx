import React, { ReactElement, useRef, useContext } from "react";
import { GroupChannel, AdminMessage, UserMessage, FileMessage } from "sendbird";
import Label, { LabelTypography, LabelColors } from "../Label";
import MessageStatus from "../MessageStatus";
import TextMessageItemBody from "../TextMessageItemBody";
import OGMessageItemBody from "../OGMessageItemBody";

import Avatar from "../../../ui/Avatar";
import ClientAdminMessage from "../../../ui/AdminMessage";
import FileMessageItemBody from "../../../ui/FileMessageItemBody";
import ThumbnailMessageItemBody from "../ThumbnailMessageItemBody";
import UnknownMessageItemBody from "../../../ui/UnknownMessageItemBody";

import { LocalizationContext } from "../../../lib/LocalizationContext";
import { OutgoingMessageStates } from "../../../utils/index";

import {
  getClassName,
  getUIKitMessageTypes,
  getUIKitMessageType,
  isTextMessage,
  isOGMessage,
  isThumbnailMessage,
  isMessageSentByMe,
  isMessageSentByOperator,
  getOutgoingMessageState,
  getSenderName,
  getMessageCreatedAt,
  isSentMessage,
  isPendingMessage,
  CoreMessageType,
} from "../../../utils";

import { isAssignmentMessage, isMaterialMessage } from '../../utils';
import AssignmentMessageItemBody from "../AssignmentMessageItemBody";
import MaterialMessageItemBody from "../MaterialMessageItemBody";
import { generateColorFromString } from "./utils";
import "./index.scss";

import MessageItemMenu from "../MessageItemMenu";

interface Props {
  chainBottom?: boolean;
  chainTop?: boolean;
  channel: GroupChannel;
  className?: string | Array<string>;
  message: AdminMessage | UserMessage | FileMessage;
  nicknamesMap?: Map<string, string>;
  userId: string;
  useReplying?: boolean;
  resendMessage?: (message: UserMessage | FileMessage) => void;
  scrollToMessage?: (createdAt: number, messageId: number) => void;
  showEdit?: (bool: boolean) => void;
  showFileViewer?: (bool: boolean) => void;
  showRemove?: (bool: boolean) => void;
  toggleReaction?: (
    message: UserMessage | FileMessage,
    reactionKey: string,
    isReacted: boolean
  ) => void;
  disabled?: boolean;
}

export default function MessageContent({
  channel,
  chainBottom = false,
  chainTop = false,
  className,
  message,
  // nicknamesMap,
  userId,
  // useReaction = false,
  // useReplying,
  // scrollToMessage,
  showEdit,
  showFileViewer,
  showRemove,
  resendMessage,
  disabled = false,
}: // showRemove,
  // toggleReaction,
  Props): ReactElement {
  const { stringSet } = useContext(LocalizationContext);
  const messageTypes = getUIKitMessageTypes();
  const avatarRef = useRef(null);


  const isByMe: boolean =
    isPendingMessage(channel, message as UserMessage | FileMessage) ||
    !isSentMessage(channel, message as UserMessage | FileMessage) ||
    isMessageSentByMe(userId, message as UserMessage | FileMessage);
  const isOperatorMessage: boolean = isMessageSentByOperator(
    message as CoreMessageType
  );

  const isByMeClassName = isByMe
    ? "rogu-message-content--outgoing"
    : "rogu-message-content--incoming";
  const chainBottomClassName = chainBottom
    ? "rogu-message-content--chain-bottom"
    : "";
  const chainTopClassName = chainTop ? "rogu-message-content--chain-top" : "";

  if (message?.isAdminMessage?.() || message?.messageType === "admin") {
    return <ClientAdminMessage message={message} />;
  }


  return (
    <div
      className={getClassName([
        className,
        "rogu-message-content",
        isByMeClassName,
        chainBottomClassName,
        chainTopClassName,
      ])}
    >
      {/* Profile picture */}
      {!isByMe && !chainTop && (
        <Avatar
          className="rogu-message-content__avatar"
          src={message?.sender?.profileUrl || ""}
          ref={avatarRef}
          height="2rem"
          width="2rem"
        />
      )}

      <div className="rogu-message-content__content">
        {/* Bubble wrapper */}
        <div className="rogu-message-content__bubble">
          <div className="rogu-message-content__bubble__header">
            {/* Sender's name */}
            {!isByMe && !chainTop && (
              <>
                <Label
                  className="rogu-message-content__sender-name"
                  color={LabelColors.ONBACKGROUND_2}
                  style={{
                    color: generateColorFromString(
                      message?.sender?.nickname || ""
                    ),
                  }}
                  type={LabelTypography.CAPTION_1}
                >
                  {getSenderName(message)}
                </Label>
                {/* Teacher label */}
                {isOperatorMessage && !chainTop && (
                  <Label
                    className="rogu-message-content__operator-label"
                    type={LabelTypography.CAPTION_3}
                  >
                    {stringSet.LABEL__OPERATOR}
                  </Label>
                )}

                <MessageItemMenu
                  className="rogu-message-content-menu__normal-menu"
                  channel={channel}
                  message={message as UserMessage | FileMessage}
                  isByMe={isByMe}
                  disabled={disabled}
                  showEdit={showEdit}
                  showRemove={showRemove}
                  resendMessage={resendMessage}
                  showFileViewer={showFileViewer}
                />
              </>
            )}
          </div>

          <div className="rogu-message-content__bubble__body">
            <div className="rogu-message-content__buble__body-text">
              {/* Message content */}
              {isTextMessage(message as UserMessage) && (
                <TextMessageItemBody
                  isByMe={isByMe}
                  message={message?.message}
                />
              )}
              {isOGMessage(message as UserMessage) && (
                <OGMessageItemBody
                  message={message as UserMessage}
                  isByMe={isByMe}
                />
              )}
              {
                isAssignmentMessage(message.customType) && (
                  <AssignmentMessageItemBody message={message as UserMessage} isByMe={isByMe} />
                )
              }
              {
                isMaterialMessage(message.customType) && (
                  <MaterialMessageItemBody message={message as UserMessage} isByMe={isByMe} />
                )
              }
              {getUIKitMessageType(message as FileMessage) ===
                messageTypes.FILE && (
                  <FileMessageItemBody
                    message={message as FileMessage}
                    isByMe={isByMe}
                  />
                )}
              {isThumbnailMessage(message as FileMessage) && (
                <>
                  <ThumbnailMessageItemBody
                    message={message as FileMessage}
                    isByMe={isByMe}
                    showFileViewer={showFileViewer}
                    isClickable={getOutgoingMessageState(channel, message) !== OutgoingMessageStates.PENDING}
                  />
                  <TextMessageItemBody
                    isByMe={isByMe}
                    mode="thumbnailCaption"
                    message={message?.name}
                  />
                </>
              )}
              {getUIKitMessageType(message as FileMessage) ===
                messageTypes.UNKNOWN && (
                  <UnknownMessageItemBody message={message} isByMe={isByMe} />
                )}
            </div>
            {
              ((!isByMe && chainTop) || isByMe) && (
                <MessageItemMenu
                  className="rogu-message-content-menu__normal-menu"
                  channel={channel}
                  message={message as UserMessage | FileMessage}
                  isByMe={isByMe}
                  disabled={disabled}
                  showEdit={showEdit}
                  showRemove={showRemove}
                  resendMessage={resendMessage}
                  showFileViewer={showFileViewer} />

              )
            }


          </div>


        </div>

        {/* Message status */}
        {!chainBottom && (
          <div className={"rogu-message-content__misc"}>
            {isByMe ? (
              <MessageStatus
                message={message}
                status={getOutgoingMessageState(channel, message)}
              />
            ) : (
              <Label
                className={"rogu-message-content__created-at"}
                type={LabelTypography.CAPTION_3}
                color={LabelColors.ONBACKGROUND_2}
              >
                {getMessageCreatedAt(message)}
              </Label>
            )}
          </div>
        )}
      </div>
    </div>
  );
}
