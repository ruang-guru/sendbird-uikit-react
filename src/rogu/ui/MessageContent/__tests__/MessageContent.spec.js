import React from "react";
import renderer from "react-test-renderer";
import { shallow } from "enzyme";

import MessageContent from "../index";

import {
  BASIC_MESSAGE_A_1,
  USER_ID_A,
} from "../../../../../__mocks__/messagesMock";

describe("MessageContent", () => {
  it("should do a snapshot test of the MessageContent DOM", function () {
    const component = renderer.create(
      <MessageContent
        userId={"random-user-id"}
        channel={{
          isGroupChannel: () => true,
          getUnreadMemberCount: (_) => 10,
          getUndeliveredMemberCount: (_) => 0,
        }}
        message={BASIC_MESSAGE_A_1}
      />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should be rendered with the classname of `.rogu-message-content`", function () {
    const component = shallow(
      <MessageContent
        userId={"random-user-id"}
        channel={{
          isGroupChannel: () => true,
          getUnreadMemberCount: (_) => 10,
          getUndeliveredMemberCount: (_) => 0,
        }}
        message={BASIC_MESSAGE_A_1}
      />
    );
    expect(component.hasClass("rogu-message-content")).toBe(true);
  });

  it("should be rendered with the className of `.rogu-message-content--incoming` to outgoing message", function () {
    const component = shallow(
      <MessageContent
        userId={"random-user-id"}
        channel={{
          isGroupChannel: () => true,
          getUnreadMemberCount: (_) => 10,
          getUndeliveredMemberCount: (_) => 0,
        }}
        message={BASIC_MESSAGE_A_1}
      />
    );
    expect(component.hasClass("rogu-message-content--incoming")).toBe(true);
  });

  it("should be rendered with the className of `.rogu-message-content--outgoing` to incoming message", function () {
    const component = shallow(
      <MessageContent
        userId={USER_ID_A}
        channel={{
          isGroupChannel: () => true,
          getUnreadMemberCount: (_) => 10,
          getUndeliveredMemberCount: (_) => 0,
        }}
        message={BASIC_MESSAGE_A_1}
      />
    );
    expect(component.hasClass("rogu-message-content--outgoing")).toBe(true);
  });

  it("should be rendered with the className of `.rogu-message-content--chain-bottom` if chainBottom is true", function () {
    const component = shallow(
      <MessageContent
        userId={"random-user-id"}
        chainBottom={true}
        channel={{
          isGroupChannel: () => true,
          getUnreadMemberCount: (_) => 10,
          getUndeliveredMemberCount: (_) => 0,
        }}
        message={BASIC_MESSAGE_A_1}
      />
    );
    expect(component.hasClass("rogu-message-content--chain-bottom")).toBe(true);
  });

  it("should be rendered with the className of `.rogu-message-content--chain-top` if chainTop is true", function () {
    const component = shallow(
      <MessageContent
        userId={"random-user-id"}
        chainTop={true}
        channel={{
          isGroupChannel: () => true,
          getUnreadMemberCount: (_) => 10,
          getUndeliveredMemberCount: (_) => 0,
        }}
        message={BASIC_MESSAGE_A_1}
      />
    );
    expect(component.hasClass("rogu-message-content--chain-top")).toBe(true);
  });
});
