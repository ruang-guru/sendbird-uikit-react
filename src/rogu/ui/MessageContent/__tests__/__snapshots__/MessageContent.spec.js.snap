// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageContent should do a snapshot test of the MessageContent DOM 1`] = `
<div
  className=" rogu-message-content rogu-message-content--incoming  "
>
  <div
    className="rogu-message-content__avatar sendbird-avatar"
    role="button"
    style={
      Object {
        "height": "2rem",
        "width": "2rem",
      }
    }
    tabIndex={0}
  >
    <div
      className="sendbird-avatar-img sendbird-image-renderer"
      style={
        Object {
          "height": "2rem",
          "maxWidth": "400px",
          "minWidth": "2rem",
          "width": "100%",
        }
      }
    >
      <div
        className="sendbird-image-renderer__image"
        style={
          Object {
            "backgroundImage": "url(https://avatars3.githubusercontent.com/u/46333979?s=460&v=4)",
            "backgroundPosition": "center",
            "backgroundRepeat": "no-repeat",
            "backgroundSize": "cover",
            "borderRadius": null,
            "height": "2rem",
            "maxWidth": "400px",
            "minWidth": "2rem",
            "position": "absolute",
            "width": "100%",
          }
        }
      />
      <img
        alt=""
        className="sendbird-image-renderer__hidden-image-loader"
        onError={[Function]}
        onLoad={[Function]}
        src="https://avatars3.githubusercontent.com/u/46333979?s=460&v=4"
      />
    </div>
  </div>
  <div
    className="rogu-message-content__content"
  >
    <div
      className="rogu-message-content__bubble"
    >
      <div
        className="rogu-message-content__bubble__header"
      >
        <span
          className="rogu-message-content__sender-name sendbird-label sendbird-label--caption-1 sendbird-label--color-onbackground-2"
          style={
            Object {
              "color": "#6073E2",
            }
          }
        >
          Kukuh Aji Sulistyo Bangun Jaya Abadi Mangkubumi
        </span>
      </div>
      <div
        className=" rogu-text-message-item-body  rogu-text-message-item-body--incoming"
      >
        <div
          className="rogu-text-message-item-body__inner"
        >
          <span
            className="rogu-text-message-item-body__message sendbird-label sendbird-label--body-1 sendbird-label--color-onbackground-1"
          >
            Welcome to the classroom!
          </span>
        </div>
      </div>
    </div>
    <div
      className="rogu-message-content__misc"
    >
      <span
        className="rogu-message-content__created-at sendbird-label sendbird-label--caption-3 sendbird-label--color-onbackground-2"
      >
        14:45
      </span>
    </div>
  </div>
</div>
`;
